docs/cookbook/prompt_engineering/chaining_based/diverse.md

```diff
@@ -0,0 +1,156 @@
+
+## Implementation
+
+Let's implement the DiVeRSe technique using Mirascope:
+
+```python
+import asyncio
+
+from pydantic import BaseModel, Field
+
+from mirascope.core import openai, prompt_template
+
+
+class PromptVariations(BaseModel):
+    variations: list[str] = Field(..., description="Variations of the original prompt")
+
+
+@openai.call(model="gpt-4o-mini", response_model=PromptVariations)
+@prompt_template(
+    """
+    Return the {num_prompts} alternate variations of the prompt which retain the
+    full meaning but uses different phrasing.
+    Prompt: {prompt}
+    """
+)
+def get_prompt_variations(prompt: str, num_prompts: int): ...
+
+
+@openai.call(model="gpt-4o-mini")
+@prompt_template(
+    """
+    Answer the following question going step by step:
+    {query}
+    """
+)
+async def zero_shot_cot(query: str): ...
+
+
+class ResponseDetails(BaseModel):
+    solution_number: int = Field(
+        ..., description="The actual number given as the answer in a solution."
+    )
+    correctness_probability: float = Field(
+        ...,
+        ge=0,
+        le=1,
+        description="An estimated probability that the given solution is correct from 0.0 to 1.0",
+    )
+
+
+@openai.call(model="gpt-4o-mini", response_model=ResponseDetails)
+@prompt_template(
+    """
+    Here is a query and a response which attempts to answer the query.
+    Prompt: {query}
+    Response: {response}
+
+    Extract the raw numerical value of the answer given by the response, and also
+    give an estimate between 0.0 and 1.0 of the probability that this solution
+    is correct.
+    """
+)
+async def evaluate_response(query: str, response: str): ...
+
+
+async def diverse(query: str, num_variations: int):
+    # Gather the variations of the prompt
+    alternate_variations = get_prompt_variations(query, num_variations - 1)
+    all_variations = alternate_variations.variations + [query]
+
+    # Generate a unique reasoning chain for each prompt variation with CoT
+    cot_tasks = [zero_shot_cot(prompt) for prompt in all_variations]
+    cot_responses = [response.content for response in await asyncio.gather(*cot_tasks)]
+
+    # Evaluate each reasoning chain
+    eval_tasks = [
+        evaluate_response(query, cot_response) for cot_response in cot_responses
+    ]
+    eval_responses = await asyncio.gather(*eval_tasks)
+
+    response_scores = {}
+    for eval_response in eval_responses:
+        if eval_response.solution_number not in response_scores:
+            response_scores[eval_response.solution_number] = 0
+        response_scores[eval_response.solution_number] += (
+            eval_response.correctness_probability
+        )
+    best_response = max(response_scores, key=response_scores.get)  # type: ignore
+    return best_response
+
+
+async def run_diverse(prompt, num_variations=3):
```

Comment:
type this to return a string and return the `response.content` of `diverse(...)`